### PR TITLE
Delete plugin settings on a successful migration of the plugin settings to cluster profile. (#8643)

### PR DESCRIPTION
**Pull request by GaneshSPatil**
Issue: #8643

Description:

* Do not delete the plugin settings when the migrate-config call fails.
* Do not delete the plugin settings when the elastic agent plugin implements
  elastic agent extension v4, which relies on plugin settings.
* Delete the plugin settings when the plugin successfully migrates the
  plugin settings to cluster profile and the plugin implements elastic
  agent extension v5 or above, which does not rely on plugin settings.

### CODE REVIEW

File: server/src/main/java/com/thoughtworks/go/server/service/ElasticAgentInformationMigratorImpl.java [hunk 1]
```diff
@@ -68,12 +73,29 @@ private boolean migrate(GoPluginDescriptor pluginDescriptor) {
...
         Plugin plugin = pluginSqlMapDao.findPlugin(pluginId);
         String pluginConfiguration = plugin.getConfiguration();
         HashMap<String, String> pluginSettings = (pluginConfiguration == null) ? new HashMap<>() : JsonHelper.fromJson(pluginConfiguration, HashMap.class);
         ReplaceElasticAgentInformationCommand command = new ReplaceElasticAgentInformationCommand(clusterProfilesService, elasticProfileService, elasticAgentExtension, pluginDescriptor, pluginSettings);
 
-        return update(command, pluginDescriptor);
+        boolean updated = update(command, pluginDescriptor);
+
+        if (updated) {
+            LOG.debug("Done Migrating elastic agent information for {} plugin", pluginId);
```
Comment (maheshp) [inline server/src/main/java/com/thoughtworks/go/server/service/ElasticAgentInformationMigratorImpl.java:85]: `Done migrating elastic agent configurations for plugin with id: '{}'`

File: server/src/main/java/com/thoughtworks/go/server/service/ElasticAgentInformationMigratorImpl.java [hunk 2]
```diff
@@ -68,12 +73,29 @@ private boolean migrate(GoPluginDescriptor pluginDescriptor) {
...
+
+            // all the plugins implementing elastic agent extension v5, does not use plugin settings and hence delete them after successful migration
+            if (!pluginManager.resolveExtensionVersion(pluginId, ELASTIC_AGENT_EXTENSION, ElasticAgentExtension.SUPPORTED_VERSIONS).equals(ElasticAgentExtensionV4.VERSION)) {
+                LOG.debug("Deleting stale plugin settings for {} plugin, after successful plugin config migration", pluginId);
+                pluginSqlMapDao.deletePluginIfExists(pluginId);
+            } else {
+                LOG.debug("Skip deleting plugin settings for {} plugin, as the plugin implements elastic agent v4 extension which uses plugin settings", pluginId);
+            }
+        } else {
+            LOG.debug("Failed migrating elastic agent information for {} plugin", pluginId);
```
Comment (maheshp) [inline server/src/main/java/com/thoughtworks/go/server/service/ElasticAgentInformationMigratorImpl.java:95]: Probably this should be logged as error in the `update ` method catch block.

File: server/src/main/java/com/thoughtworks/go/server/dao/PluginSqlMapDao.java [hunk 3]
```diff
@@ -113,4 +113,23 @@ protected void doInTransactionWithoutResult(TransactionStatus status) {
...
+        String cacheKey = cacheKeyForPluginSettings(pluginId);
+        Plugin plugin = this.findPlugin(pluginId);
+
+        if (plugin instanceof NullPlugin) {
+            return;
+        }
+
+        synchronized (cacheKey) {
+            transactionTemplate.execute((TransactionCallback) transactionStatus -> sessionFactory.getCurrentSession()
+                    .createQuery(String.format("delete from %s where pluginId=:pluginId", Plugin.class.getSimpleName()))
```
Comment (maheshp) [inline server/src/main/java/com/thoughtworks/go/server/dao/PluginSqlMapDao.java:128]: Couldn't we just do `sessionFactory.getCurrentSession().delete(plugin);`

Comment (GaneshSPatil) [reply server/src/main/java/com/thoughtworks/go/server/dao/PluginSqlMapDao.java:128]: `transactionTemplate.execute` method expects a return value and `sessionFactory.getCurrentSession().delete` returns a void. This requires adding a new execute method which supports void return type.

**Review by maheshp — APPROVED** [review]

**Pull request merged**